sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala

```diff
@@ -230,12 +256,29 @@ case class InsertIntoHiveTable(
         var doHiveOverwrite = overwrite
 
         if (oldPart.isEmpty || !ifPartitionNotExists) {
+          // SPARK-29295: When insert overwrite to a Hive external table partition, if the
+          // partition does not exist, Hive will not check if the external partition directory
+          // exists or not before copying files. So if users drop the partition, and then do
+          // insert overwrite to the same partition, the partition will have both old and new
+          // data.
+          val updatedPart = if (oldPart.isEmpty && overwrite
+              && table.tableType == CatalogTableType.EXTERNAL) {
+            AlterTableAddPartitionCommand(
+              table.identifier, Seq((partitionSpec, None)), ifNotExists = true).run(sparkSession)
+            externalCatalog.getPartitionOption(
```

Comment:
I feel it's better to construct the partition directory and remove it, instead of sending 2 metadata requests.

Reply:
This is also an option. Metadata request was just used to construct partition directory for us.

I changed to your suggestion.